File: backend/vision/app/vms/common/media_token.py
"""Short-lived signed **media token** — the live-stream hot-path credential.

vision mints one of these per playback session; the browser carries it as
``?token=<t>`` on every HLS segment / WHEP request, and Traefik ForwardAuth (P2-C)
calls ``GET /api/v1/vms/media/verify`` which validates it here before letting the
request reach MediaMTX. Because it gates EVERY segment it is:

  * **Stateless** — a plain HS256 JWT signed with the SAME kernel ``jwt_secret``
    the access tokens use (no new secret, no DB hit to verify). Distinguished from
    an access token by ``sub_type="media"`` (access tokens use ``type="access"``),
    so a media token can NEVER be used as an API token and vice-versa.
  * **Short-lived** — TTL ``VE_MEDIA_TOKEN_TTL_SEC`` (default 300s), renewable via
    the session ``/renew`` endpoint so long live views don't drop.

Claims: ``{sub_type:"media", tenant_id, camera_id, session_id, iat, exp}``.

Kept dependency-light (pyjwt only, already a kernel dep) and independent of the
kernel auth module so it can be imported on the hot path without pulling FastAPI
security wiring.
"""

from __future__ import annotations

import hashlib
import os
import time

import jwt

_ALG = "HS256"
_SUB_TYPE = "media"
_DEFAULT_TTL = 300


def media_token_ttl() -> int:
    """TTL (seconds) for a minted media token — ``VE_MEDIA_TOKEN_TTL_SEC`` (default 300)."""
    raw = os.environ.get("VE_MEDIA_TOKEN_TTL_SEC", "").strip()
    try:
        ttl = int(raw) if raw else _DEFAULT_TTL
    except ValueError:
        ttl = _DEFAULT_TTL
    return ttl if ttl > 0 else _DEFAULT_TTL


def _secret() -> str:
    # Lazy import avoids an import cycle + keeps the module cheap to import.
    from kernel.config import get_settings

    return get_settings().jwt_secret


def mint_media_token(
    *,
    tenant_id: str | None,
    camera_id: str,
    session_id: str,
    ttl_seconds: int | None = None,
    mode: str = "live",
) -> tuple[str, int]:
    """Mint a media token → ``(token, exp_epoch_seconds)``.

    ``tenant_id`` is stringified (or the reserved ``"platform"`` for a NULL-tenant /
    super-admin session) so the verify path can compare it to the camera's tenant.

    ``mode`` (``"live"`` default, ``"playback"`` for P4-A recorded playback) is
    carried as a claim so a token can be audited/scoped by what it gates. It does NOT
    change the signature/type — a ``mode:playback`` token is a normal media token and
    ``verify_media_token`` accepts it identically (it gates the same MediaMTX proxy).
    """
    ttl = ttl_seconds if (ttl_seconds and ttl_seconds > 0) else media_token_ttl()
    now = int(time.time())
    exp = now + ttl
    claims = {
        "sub_type": _SUB_TYPE,
        "tenant_id": tenant_id if tenant_id is not None else "platform",
        "camera_id": camera_id,
        "session_id": session_id,
        "mode": mode or "live",
        "iat": now,
        "exp": exp,
    }
    token = jwt.encode(claims, _secret(), algorithm=_ALG)
    # pyjwt<2 returns bytes; normalise to str.
    if isinstance(token, bytes):  # pragma: no cover - pyjwt>=2 returns str
        token = token.decode("utf-8")
    return token, exp


def verify_media_token(token: str) -> dict:
    """Decode + verify a media token → its claims dict.

    Raises ``jwt.PyJWTError`` (or ``ValueError``) on any signature/expiry/type
    problem — the caller maps that to a 401. Fast + stateless: a single HMAC verify,
    no DB.
    """
    if not token:
        raise ValueError("missing media token")
    payload = jwt.decode(token, _secret(), algorithms=[_ALG])
    if payload.get("sub_type") != _SUB_TYPE:
        raise ValueError("not a media token")
    if not payload.get("camera_id") or not payload.get("session_id"):
        raise ValueError("media token missing camera/session")
    return payload


def token_hash(token: str) -> str:
    """SHA-256 hex of a token — what we persist at rest (never the raw token)."""
    return hashlib.sha256(token.encode("utf-8")).hexdigest()

File: backend/nvr/internal/mediatoken/mediatoken.go
// Package mediatoken mints and verifies the node's own short-lived **media
// token** — the live/playback hot-path credential (spec §5.5). It is the Go twin
// of vision's vms/common/media_token.py: an HS256 JWT signed with the node's
// anchor secret (VE_JWT_SECRET — enrollment-shared when the node is enrolled,
// node-local when standalone), carrying the SAME claim shape so a token minted by
// either side verifies on the other.
//
// The browser carries the token as ?token=<t> on every HLS segment / WHEP
// request; the MediaMTX ForwardAuth (GET /api/v1/nvr/media/verify) validates it
// with a single HMAC — no DB hit — before letting the request reach the media
// plane. Distinguished from an access token by sub_type="media" so it can never
// be replayed as an API token (localauth.Authenticate rejects it) and vice-versa.
package mediatoken

import (
	stderrors "errors"
	"os"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

const (
	// alg is HS256 — matches the Python core + vision media_token.py.
	alg = "HS256"
	// subTypeMedia marks a media (streaming) token, distinct from an access token.
	subTypeMedia = "media"
	// defaultTTL is the mint TTL when the caller passes ttl<=0 (matches vision's
	// VE_MEDIA_TOKEN_TTL_SEC default of 300s).
	defaultTTL = 300 * time.Second
)

// ErrInvalid is returned by Verify for any signature/expiry/type problem — the
// public route maps it to a 401/403.
var ErrInvalid = stderrors.New("invalid media token")

// Claims is the media token payload — identical field set + JSON names to
// vision/app/vms/common/media_token.py ({sub_type, tenant_id, camera_id,
// session_id, mode, iat, exp}). RegisteredClaims supplies iat/exp so jwt handles
// expiry validation on Verify.
type Claims struct {
	SubType   string `json:"sub_type"`
	TenantID  string `json:"tenant_id"`
	CameraID  string `json:"camera_id"`
	SessionID string `json:"session_id"`
	Mode      string `json:"mode"`
	jwt.RegisteredClaims
}

// secret returns the node's anchor HMAC key (VE_JWT_SECRET). Read on every
// mint/verify so a key rotation via env restart takes effect; falls back to the
// same default gokernel config uses so a dev boot without the env still works.
func secret() []byte {
	if v := os.Getenv("VE_JWT_SECRET"); v != "" {
		return []byte(v)
	}
	return []byte("change-me-in-prod")
}

// Mint issues a media token for a camera → (token, exp). tenantID is stringified
// as-is; pass "platform" for a NULL-tenant / super-admin session (mirrors vision).
// mode is "live" (default) or "playback"; it is carried as a claim for audit/scope
// but does NOT change the signature — Verify accepts either identically. ttl<=0
// uses the default (300s).
func Mint(cameraID, tenantID, sessionID, mode string, ttl time.Duration) (string, time.Time, error) {
	if ttl <= 0 {
		ttl = defaultTTL
	}
	if mode == "" {
		mode = "live"
	}
	if tenantID == "" {
		tenantID = "platform"
	}
	now := time.Now()
	exp := now.Add(ttl)
	claims := Claims{
		SubType:   subTypeMedia,
		TenantID:  tenantID,
		CameraID:  cameraID,
		SessionID: sessionID,
		Mode:      mode,
		RegisteredClaims: jwt.RegisteredClaims{
			IssuedAt:  jwt.NewNumericDate(now),
			ExpiresAt: jwt.NewNumericDate(exp),
		},
	}
	token, err := jwt.NewWithClaims(jwt.SigningMethodHS256, claims).SignedString(secret())
	if err != nil {
		return "", time.Time{}, err
	}
	return token, exp, nil
}

// Verify validates a media token and returns its claims. Fast + stateless: a
// single HMAC verify (no DB). Any signature/expiry/type problem → ErrInvalid.
func Verify(token string) (*Claims, error) {
	if token == "" {
		return nil, ErrInvalid
	}
	claims := &Claims{}
	parsed, err := jwt.ParseWithClaims(token, claims, func(t *jwt.Token) (any, error) {
		if _, ok := t.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, ErrInvalid
		}
		return secret(), nil
	}, jwt.WithValidMethods([]string{alg}))
	if err != nil || !parsed.Valid {
		return nil, ErrInvalid
	}
	if claims.SubType != subTypeMedia {
		return nil, ErrInvalid
	}
	if claims.CameraID == "" || claims.SessionID == "" {
		return nil, ErrInvalid
	}
	return claims, nil
}
